### PR TITLE
Relay support

### PR DESCRIPTION
**Pull request by tillkruss**
This is a work in progress.

## TODOs

- [x] Deprecate phpiredis + webdis now
- [x] Pass most Redis Stack tests
- [x] Replace `phpiredis` instructions with Relay instructions
- [x] Pass most tests (except `relay-incompatible` group)
- [x] Remove all `relay-*` groups except `relay-incompatible`
- [ ] Add `examples/relay_events.php` 
- [ ] Finish [Wiki page](https://github.com/predis/predis/wiki)
- [ ] Make `onInvalidated()` and other methods easier to access

## Followup
- Support `compression` and `serializer` options and handle `Predis\Command\RawCommand()` calls 
- Review `relay-fixme` test group

## Tests

```bash
./vendor/bin/phpunit -c phpunit.relay.xml --exclude-group slow,relay-incompatible
```

### CODE REVIEW

**Comment by szepeviktor** [comment]
@tillkruss Please consider not naming that branch like a tag.
It could be `3.x` or something ...

**Comment by szepeviktor** [comment]
> how can we rename

Maybe keep both old and new until you merge all 5 other PR-s.

**Comment by szepeviktor** [comment]
This seems unsporrted

> ✗ relay Could not install relay on PHP 7.2.34

**Comment by szepeviktor** [comment]
> 2780 Segmentation fault      (core dumped)

It this message from Relay?

**Comment by tillkruss** [comment]
> > 2780 Segmentation fault      (core dumped)
> 
> It this message from Relay?

No, that’s PHP crashing. This is a WIP, no need to review it.

**Review by vladvildanov — CHANGES_REQUESTED** [review]

File: src/Command/Redis/COMMAND.php [hunk 1]
```diff
@@ -26,4 +26,15 @@ public function getId()
     {
         return 'COMMAND';
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function parseResponse($data)
+    {
+        // Relay (RESP3) uses maps and it might be good
```
[the remark this answers is not in the document]
Comment (tillkruss) [reply src/Command/Redis/COMMAND.php:35]: Relay is RESP3 only, but Predis itself can remain RESP2 for now 👍

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -57,14 +58,18 @@ jobs:
...
+      - name: Run tests
         if: ${{ matrix.php != '8.1' || matrix.redis != '7' }}
         run: vendor/bin/phpunit --verbose
 
-      - name: Run PHPUnit tests with coverage
+      - name: Run tests with coverage
         if: ${{ matrix.php == '8.1' && matrix.redis == '7' }}
         run: vendor/bin/phpunit --verbose --coverage-clover build/logs/clover.xml --coverage-filter ./src
 
+      - name: Run tests using Relay
```
[the remark this answers is not in the document]
Comment (tillkruss) [reply .github/workflows/tests.yml:69]: The first step runs all Predis tests, the 2nd step runs to gather code coverage. 

The 3rd step I added runs all unit tests using Relay to ensure it acts just like Predis expects.

File: src/Transaction/MultiExec.php [hunk 3]
```diff
@@ -375,7 +378,8 @@ public function execute($callable = null)
 
             $execResponse = $this->call('EXEC');
 
-            if ($execResponse === null) {
+            // The additional `false` check is needed for Relay, let's hope it won't break anything
+            if ($execResponse === null || $execResponse === false) {
```
[the remark this answers is not in the document]
Comment (tillkruss) [reply src/Transaction/MultiExec.php:382]: I rather leave it strict for now, `empty()` has too many falsey cases.

File: src/Command/Redis/COMMAND.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (vladvildanov) [reply src/Command/Redis/COMMAND.php:35]: @chayim Adding RESP3 support is also a near future change that should be done. It looks like to cover RESP3 needs it's more changes requires, right now we have a lot of discussion around it.

[the remark this answers is not in the document]
Comment (tillkruss) [reply src/Command/Redis/COMMAND.php:35]: Is that something you want to squeeze into v3, or into a future release?

**Pull request closed without merging**